core/modules/replica/python/schema_migration.py

```diff
@@ -0,0 +1,219 @@
+        -------
+        Integer number
+        """
+
+        # Initial qservReplica implementation did not have version number stored at all,
+        # and we call this version 0. Since version=1 version number is stored in
+        # QMetadata table with key="version"
+        if not utils.tableExists(self.engine, "QMetadata"):
+            _log.debug("QMetadata missing: version=0")
+            return 0
+        else:
+            query = "SELECT value FROM QMetadata WHERE metakey = 'version'"
+            result = self.engine.execute(query)
+            row = result.first()
+            if row:
+                _log.debug("found version in database: %s", row[0])
+                return int(row[0])
+
+    def latest_version(self):
+        """Returns latest known schema version.
+
+        Returns
+        -------
+        Integer number
+        """
+        if self.scripts:
+            version = max(script[1] for script in self.scripts)
+            _log.debug("latest version from migration scripts: %s", version)
+        else:
+            # no migration scripts - current version is latest
+            version = self.current_version()
+            _log.debug("no migration scripts, returning current version %s", version)
+        return version
+
+    def migrations(self):
+        """Returns all known migrations.
+
+        Returns
+        -------
+        List of tuples, each tuple contains three elements:
+        - starting schema version (int)
+        - final schema version (usually starting + 1)
+        - migration script name (or any arbitrary string)
+        """
+        return [script[:3] for script in self.scripts]
+
+    def migrate(self, version=None, do_migrate=False):
+        """Perform schema migration from current version to given version.
+
+        Parameters
+        ----------
+        version : int or None
+            If None then migrate to latest known version, otherwise only
+            migrate to given version.
+        do_migrate : bool
+            If True performa migration, otherwise only print steps that
+            should be performed
+
+        Returns
+        -------
+        None if no migrations were performed or the version number at
+        which migration has stopped.
+
+        Raises
+        ------
+        Exception is raised for any migration errors. The state of the
+        database is not guaranteed to be consistent after exception.
+        """
+
+        # checks for requested version
+        if version is not None:
+            current = self.current_version()
+            if current >= version:
+                _log.debug("current version (%s) is already same or newer as requested (%s)",
+                           current, version)
+                return None
+            final_versions = [s[1] for s in self.scripts]
+            if version not in final_versions:
+                raise ValueError("No known migration scripts for requested version ({})".format(version))
```

Comment:
f strings are kind of cool: `f"No known migration scripts for requested version ({version})"`